### ISSUE TITLE
Add MySQL-Support

### ISSUE THREAD

**Issue by krombel** [matrix-org/synapse#2112]
I would like to use a MySQL-Database as backend for synapse.

In #123 I can see that someone worked on the support for MySQL but switched then to PostgreSQL.

What was the reason for that? May it be possible to add support for synapse now?

Thank you

**Comment by ChazyTheBest** [comment]
I would also like MySQL support.

**Comment by ara4n** [comment]
https://matrix.org/jira/browse/SYN-404 has the details of why synapse hasn't got mysql support, and why we don't have any plans to implement it any time soon.

We can't stop other people from trying though :)

**Comment by MurzNN** [comment]
Seems that the blocker is:

> Synapse requires the ability to store full UTF8, which means using the MySQL charset utf8mb4 (not utf8, which only works for 3 byte utf8). MySQL only supports binary or case insensitive collations for utf8mb4, so we are forced to use binary collations.
 
But new stable MySQL 5.5 version have full support of `utf8mb4` collation, not only binary, but `utf8mb4_unciode_ci` and other ways. Does now we can implement MySQL backend support in Synapse?

**Comment by ShellCode33** [comment]
It'd be great indeed :) I'm also using MySQL and forced to use sqlite which I'm pretty sure slows down performance

**Comment by ara4n** [comment]
If someone wanted to go through porting the schema & queries to mysql they'd be welcome to. however, synapse does not have a backend abstraction built into it currently, and i strongly suspect this would be a lot of fragile work without much benefit. we certainly wouldn't be able to support it from the core team.

**Comment by Protected** [comment]
It's useful when you have existing mysql/mariadb infrastructure that you could use but no postgres infrastructure.

**Comment by ptman** [comment]
what is SYN-404 by github issue id?

**Comment by aggarwaldev** [comment]
@ara4n Can we please reconsider support for mysql? 
mysql 8 has brought massive performance boost to the DB.

**Comment by V1TA5** [comment]
+1

**Comment by bmiller59** [comment]
I recommend supporting mysql by implementing the backend abstraction / ORM approach @ara4n mentioned earlier. A backend abstraction would open the door to other data backends in the future, which I think will be desirable.

**Comment by menturion** [comment]
+1

**Comment by mandrillmaze** [comment]
+1

**Comment by ptman** [comment]
@menturion the layer will be available quicker if someone writes it

And please, no +1 comments, use reactions instead

**Comment by bmiller59** [comment]
@ptman Could someone who is familiar with the code base provide a quick high level outline of a strategy to efficiently implement a backend data layer abstraction? That might inspire someone to take up the challenge.

**Comment by pmdevita** [comment]
I'm potentially interested in taking this up, can someone give me an overview of what would need to be done?

**Comment by FreeManNumOne** [comment]
@pmdevita Are you adding mysql storage?

**Comment by richvdh** [comment]
I'm going to close this. It's unlikely we'd accept PRs to add mysql support to synapse due to the maintenance overhead this would add.

**Comment by army1349** [comment]
> 
> 
> I'm going to close this. It's unlikely we'd accept PRs to add mysql support to synapse due to the maintenance overhead this would add.

Is there now maintenance overhead because of SQLite+PostgreSQL?

**Comment by ptman** [comment]
AFAIK sqlite doesn't have strong typing and postgresql is closer to sqlite than mysql.

**Comment by richvdh** [comment]
> Is there now maintenance overhead because of SQLite+PostgreSQL?

yes, and it's significant. adding a third database would make the situation even worse.

**Comment by army1349** [comment]
I took a look and now I understand. This was never designed for multiple databases.

**Comment by chpatton013** [comment]
> yes, and it's significant. adding a third database would make the situation even worse.

You really should reconsider this position. Putting in the leg-work to use an abstraction layer for database access will solve this problem for all future implementations, not just the one you like at the moment.

**Comment by karolyi** [comment]
SQLAlchemy might be the perfect tool to abstract off the postgresql-specific queries into an actual ORM that is DB-independent:

https://www.sqlalchemy.org/features.html

**Comment by Sethox** [comment]
> > yes, and it's significant. adding a third database would make the situation even worse.
> 
> You really should reconsider this position. Putting in the leg-work to use an abstraction layer for database access will solve this problem for all future implementations, not just the one you like at the moment.

Yeah, agreeable to the max. The irony is, Matrix-synapse is suppose to be a connection hub for other components/modules/services and apparently databases are not one of them.

I guess what the developers are saying is "provide with code" without actually saying it (being open source and all).

**Comment by Cyborgscode** [comment]
Adding Mysql will increase the adoption rate, as postgresql is not so common in webserver configuration as mysql is. 
postgresql is terrible to administrate from MPOV as a hoster.

**Comment by ptman** [comment]
@Cyborgscode synapse is probably quite hard to run on a generic LAMP hosting service. Postgres isn't the hardest requirement.

**Comment by Cyborgscode** [comment]
Not at all. Besides the postgresql part, it was install, configure and run.. Pretty easy.

**Comment by richvdh** [comment]
Folks, I'm going to lock this, because I don't think further debate is productive.

We hear the feedback, and have sympathy with those who are more comfortable with MySQL / MariaDB for various reasons; however I think we've made our position on this plain.

As noted, Synapse does not use a database abstraction layer. There are advantages and disadvantages to doing so, but Synapse is a database-heavy application, and the current approach has meant that we have been able to hand-optimise queries and use advanced Postgres features in a way which no abstraction layer would have allowed. We could maybe bypass the abstraction layer for certain queries, but in doing so we lose many of the advantages of having the abstraction layer in the first place.

Every feature comes at an opportunity cost, in that developing and maintaining that feature takes time which is not available to spend on other features/bugfixes.  Supporting a new database engine is no different: each dialect of SQL brings overhead, not only in terms of initial development, but also ongoing in maintenance, testing and support. A DB abstraction layer helps to reduce that overhead, but cannot fully solve it.

In summary, we as the Synapse Core team feel that our time is better spent on other things such as improving Synapse's performance and reliability than on supporting MySQL.
